scipy/_lib/_util.py

```diff
@@ -1111,11 +1111,38 @@ def _dict_formatter(d, n=0, mplus=1, sorter=None):
 "core" shapes. However, array argument(s) of this function may have additional
 "batch" dimensions prepended to the core shape. In this case, the array is treated
 as a batch of lower-dimensional slices; see :ref:`linalg_batch` for details.
-Note that calls with zero-size batches are unsupported and will raise a ``ValueError``.
 """
 
 
-def _apply_over_batch(*argdefs):
+def output_from_signature(arrays, batch_shape, core_shapes, signature):
+    dtype = np.result_type(*arrays)
+
+    # parse more efficiently with regex?
```

Comment:
This transforms the NEP 5 `signature` to a dictionary.
- the key is a tuple `(i, j)` where:
 - `i` is the index of the argument in the `arrays` tuple
 - `j` is the index of an element of the argument's shape
- the value is a (one-character) `str` used to identify the length along the axis

For instance, the signature string `'(m, n),(p)'` would produce the dictionary `{(0, 0): 'm', (0, 1): 'n', (1, 0): 'p'}`. 

I haven't come back to refactor this. There might be much more efficient solution, perhaps using a different data structure. Whatever works.